tests/e2e/tkgs_ha.go

```diff
@@ -2395,4 +2395,374 @@ var _ = ginkgo.Describe("[csi-tkgs-ha] Tkgs-HA-SanityTests", func() {
+		8. Validate the CNS metadata
+		9. Scale down the sts to 0
+		10.Delete Statefullset,PVC,POD,SC
+	*/
+	ginkgo.It("Validate statefullset creation with  POD affinity and POD Anti affinity", func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ginkgo.By("CNS_TEST: Running for GC setup")
+
+		cleanupsts := false
+		nodeList, _ := fnodes.GetReadySchedulableNodes(client)
+
+		ginkgo.By("Create statefulset with parallel pod management policy with replica 3")
+		createResourceQuota(client, namespace, rqLimit, zonalWffcPolicy)
+		scParameters[svStorageClassName] = zonalWffcPolicy
+		storageclass, err := client.StorageV1().StorageClasses().Get(ctx, zonalWffcPolicy, metav1.GetOptions{})
+		if !apierrors.IsNotFound(err) {
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+
+		// Creating StatefulSet service
+		ginkgo.By("Creating service")
+		service := CreateService(namespace, client)
+		defer func() {
+			deleteService(namespace, client, service)
+		}()
+
+		statefulset := GetStatefulSetFromManifest(namespace)
+		ginkgo.By("Creating statefulset with POD Anti affinity")
+		allowedTopologies := getTopologySelector(allowedTopologyHAMap, categories,
+			tkgshaTopologyLevels)
+		framework.Logf("allowedTopo: %v", allowedTopologies)
+		statefulset.Spec.PodManagementPolicy = appsv1.ParallelPodManagement
+		statefulset.Spec.VolumeClaimTemplates[len(statefulset.Spec.VolumeClaimTemplates)-1].
+			Annotations["volume.beta.kubernetes.io/storage-class"] = storageclass.Name
+		statefulset.Spec.Template.Spec.Affinity = new(v1.Affinity)
+		statefulset.Spec.Template.Spec.Affinity.NodeAffinity = new(v1.NodeAffinity)
+		statefulset.Spec.Template.Spec.Affinity.NodeAffinity.
+			RequiredDuringSchedulingIgnoredDuringExecution = new(v1.NodeSelector)
+		statefulset.Spec.Template.Spec.Affinity.NodeAffinity.
+			RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = getNodeSelectorTerms(allowedTopologies)
+
+		statefulset.Spec.Template.Spec.Affinity.PodAntiAffinity = new(v1.PodAntiAffinity)
+		statefulset.Spec.Template.Spec.Affinity.PodAntiAffinity.
+			RequiredDuringSchedulingIgnoredDuringExecution = getPodAffinityTerm(allowedTopologyHAMap)
+
+		*statefulset.Spec.Replicas = 3
+		framework.Logf("Statefulset spec: %v", statefulset)
+		ginkgo.By("Create Statefullset with PodAntiAffinity")
+		cleanupsts = true
+		CreateStatefulSet(namespace, statefulset, client)
+		replicas := *(statefulset.Spec.Replicas)
+
+		defer func() {
+			if cleanupsts {
+				framework.Logf("cleaning up statefullset with podAtiAffinity")
+				cleaupStatefullset(client, ctx, namespace, statefulset)
+			}
+		}()
+
+		ginkgo.By("Verify annotations on SVC PV and required node affinity details on SVC PV and GC PV")
+		ginkgo.By("Verify pod gets scheduled on appropriate nodes preset in the availability zone")
+		verifyStsVolumeMetadata(client, ctx, namespace, statefulset, replicas,
+			allowedTopologyHAMap, categories, zonalPolicy, nodeList, f)
```

Comment:
Does this function validate the pod's availability zone also?

Reply:
Yes , This verifies node affinity and pod location of volumes on correct zones